be/src/service/service_be/lake_service.cpp

```diff
@@ -4,27 +4,146 @@
+Status apply_txn_log(const lake::TxnLog& log, lake::TabletMetadata* metadata) {
+    // apply txn log
+    if (log.has_op_write()) {
+        if (log.op_write().has_rowset() && log.op_write().rowset().segments_size() > 0) {
+            auto rowset = metadata->add_rowsets();
+            rowset->CopyFrom(log.op_write().rowset());
+            rowset->set_id(metadata->next_rowset_id());
+            metadata->set_next_rowset_id(metadata->next_rowset_id() + std::max<int>(1, rowset->segments_size()));
+        }
+    }
+
+    if (log.has_op_compaction()) {
+        return Status::NotSupported("does not support apply compaction log yet");
+    }
+
+    if (log.has_op_schema_change()) {
+        return Status::NotSupported("does not support apply schema change log yet");
+    }
+
+    return Status::OK();
+}
+
 void LakeServiceImpl::publish_version(::google::protobuf::RpcController* controller,
                                       const ::starrocks::lake::PublishVersionRequest* request,
                                       ::starrocks::lake::PublishVersionResponse* response,
                                       ::google::protobuf::Closure* done) {
     brpc::ClosureGuard guard(done);
     (void)controller;
-    (void)request;
-    response->mutable_status()->set_status_code(TStatusCode::NOT_IMPLEMENTED_ERROR);
+    if (!request->has_version()) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing version");
+        return;
+    }
+    if (request->txn_ids_size() == 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing txn_ids");
+        return;
+    }
+    if (request->tablet_ids_size() == 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing tablet_ids");
+        return;
+    }
+    int64_t base_version = request->version() - request->txn_ids_size();
+    if (base_version < 1) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("invalid version");
+        return;
+    }
+    // Will not update status code since here, only failed_tablets will be updated.
+    response->mutable_status()->set_status_code(TStatusCode::OK);
+
+    // TODO move the execution to TaskWorkerPool
+    for (const auto& tablet_id : request->tablet_ids()) {
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(WARNING) << "Fail to get tablet " << tablet_id << ": " << tablet.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+        auto base_metadata = tablet->get_metadata(base_version);
+        if (base_metadata.status().is_not_found() && tablet->get_metadata(request->version()).ok()) {
```

Comment:
Why check base tablet metadata `is_not_found`?

Reply:
The GC module may have deleted the old version metadata.

Reply:
comment updated, PTAL